internal/multistorage/cache/file.go

```diff
@@ -0,0 +1,95 @@
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+// StatusFile is stored on a disk and therefore shared between all WAL-G processes and commands.
+const StatusFile = ".walg_storage_status_cache"
+
+type storageStatuses map[string]status
+
+type status struct {
+	Alive   bool      `json:"alive"`
+	Checked time.Time `json:"checked"`
+}
+
+func updateFileContent(oldContent storageStatuses, checkResult map[string]bool) (newContent storageStatuses) {
+	newContent = make(storageStatuses, len(oldContent))
+	for storage, status := range oldContent {
+		newContent[storage] = status
+	}
+
+	checkTime := time.Now()
+	for s, alive := range checkResult {
+		newContent[s] = status{
+			Alive:   alive,
+			Checked: checkTime,
+		}
+	}
+
+	return newContent
+}
+
+func readFile() (storageStatuses, error) {
+	file, err := os.OpenFile(StatusFile, os.O_RDONLY, 0666)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("open cache file: %w", err)
+	}
+	defer func() { _ = file.Close() }()
+
+	err = unix.Flock(int(file.Fd()), unix.LOCK_SH)
```

Comment:
In the previous implementation, the cache file wasn't locked. In some improbable coincidences, it might lead to a problem with content corruption, because there may be multiple wal-g processes writing to this file at the same time. That's why I added locking here.

To be honest, it's not too critical, because the problem mustn't exist if just a few storages are used and the whole file fits in 512 bytes or something (disk block size or the minimum writable data size declared by OS)